internal/metrics/otelmetrics/factory_test.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/jaegertracing/jaeger/pkg/metrics"
+	"github.com/jaegertracing/jaeger/pkg/testutils"
+)
+
+func TestMain(m *testing.M) {
+	testutils.VerifyGoLeaks(m)
+}
+
+func newTestFactory(t *testing.T, registry *promReg.Registry) metrics.Factory {
+	exporter, err := prometheus.New(prometheus.WithRegisterer(registry))
+	require.NoError(t, err)
+	meterProvider := sdkmetric.NewMeterProvider(sdkmetric.WithReader(exporter))
+	return otelmetrics.NewFactory(meterProvider)
+}
+
+func findMetric(t *testing.T, registry *promReg.Registry, name string) *promModel.MetricFamily {
+	metricFamilies, err := registry.Gather()
+	require.NoError(t, err)
+
+	for _, mf := range metricFamilies {
+		t.Log(mf.GetName())
+		if mf.GetName() == name {
+			return mf
+		}
+	}
+	require.NotNil(t, name, "Expected to find Metric Family")
+	return nil
+}
+
+func TestCounter(t *testing.T) {
+	registry := promReg.NewPedanticRegistry()
+	factory := newTestFactory(t, registry)
+	counter := factory.Counter(metrics.Options{
+		Name: "test_counter",
+		Tags: map[string]string{"tag1": "value1"},
+	})
+	require.NotNil(t, counter)
+	counter.Inc(1)
+	counter.Inc(1)
+
+	// Find the metric in the registry
+	testCounter := findMetric(t, registry, "test_counter_total")
+
+	// Check the metric values
+	metrics := testCounter.GetMetric()
+	assert.Equal(t, float64(2), metrics[0].GetCounter().GetValue())
+}
+
+func TestGauge(t *testing.T) {
+	registry := promReg.NewPedanticRegistry()
+	factory := newTestFactory(t, registry)
+	gauge := factory.Gauge(metrics.Options{
+		Name: "test_gauge",
+		Tags: map[string]string{"tag1": "value1"},
+	})
+	require.NotNil(t, gauge)
+	gauge.Update(2)
+
+	testGauge := findMetric(t, registry, "test_gauge")
+
+	metrics := testGauge.GetMetric()
+	assert.Equal(t, float64(2), metrics[0].GetGauge().GetValue())
+}
+
+func TestHistogram(t *testing.T) {
+	registry := promReg.NewPedanticRegistry()
+	factory := newTestFactory(t, registry)
+	histogram := factory.Histogram(metrics.HistogramOptions{
+		Name: "test_histogram",
+		Tags: map[string]string{"tag1": "value1"},
+	})
+	require.NotNil(t, histogram)
+	histogram.Record(1.0)
+
+	testHistogram := findMetric(t, registry, "test_histogram")
+
+	metrics := testHistogram.GetMetric()
+	assert.Equal(t, float64(1), metrics[0].GetHistogram().GetSampleSum())
+}
+
+func TestTimer(t *testing.T) {
+	registry := promReg.NewPedanticRegistry()
+	factory := newTestFactory(t, registry)
+	timer := factory.Timer(metrics.TimerOptions{
+		Name: "test_timer",
+		Tags: map[string]string{"tag1": "value1"},
+	})
+	require.NotNil(t, timer)
+	timer.Record(100 * time.Millisecond)
+
+	testTimer := findMetric(t, registry, "test_timer_seconds")
+
+	metrics := testTimer.GetMetric()
+	assert.Equal(t, float64(0.1), metrics[0].GetHistogram().GetSampleSum())
+}
+
+func TestNamespace(t *testing.T) {
+	registry := promReg.NewPedanticRegistry()
+	factory := newTestFactory(t, registry)
+	nsFactory := factory.Namespace(metrics.NSOptions{
```

Comment:
there are more permutations that this function supports, you should test them all via a table driven test